### PR TITLE
Nn 5119 create punishment comment

### PR DESCRIPTION
**Pull request by p-nik**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/ReportedAdjudicationDto.kt [hunk 1]
```diff
@@ -259,3 +261,15 @@ data class SuspendedPunishmentDto(
...
   val punishment: PunishmentDto,
 )
+
+@Schema(description = "punishment comment")
+data class PunishmentCommentDto(
+  @Schema(description = "punishment comment id used for edit and delete")
+  val id: Long? = null,
+  @Schema(description = "comment on punishment")
+  val comment: String,
+  @Schema(description = "username of the person created or updated the comment")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/ReportedAdjudicationDto.kt:271]: they mandatory, remove elvis ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/ReportedAdjudicationDto.kt [hunk 2]
```diff
@@ -259,3 +261,15 @@ data class SuspendedPunishmentDto(
...
+  val createdByUserId: String? = null,
+  @Schema(description = "date and time comment was created or updated")
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/dtos/ReportedAdjudicationDto.kt:273]: same here.  no need for ? as its mandatory

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt [hunk 3]
```diff
@@ -351,6 +353,27 @@ class PunishmentsService(
...
 
+  fun createPunishmentComment(
+    adjudicationNumber: Long,
+    punishmentComment: PunishmentCommentRequest,
+  ): ReportedAdjudicationDto {
+    val reportedAdjudication = findByAdjudicationNumber(adjudicationNumber).also {
+      it.validateCanAddPunishmentComment()
+    }
+
+    if (punishmentComment.comment.isBlank()) {
```
Comment (timoj58) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt:364]: good spot btw.  But lets not fix this here.  Raise a ticket for front end to trim all text before sending, as this issue is present across the system currently.  Happy for you to remove this check.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt [hunk 4]
```diff
@@ -401,5 +424,11 @@ class PunishmentsService(
         throw ValidationException("outcome is a caution - no further punishments can be added")
       }
     }
+
+    fun ReportedAdjudication.validateCanAddPunishmentComment() {
+      if (this.getPunishments().isEmpty()) {
+        throw EntityNotFoundException("Punishments not found for adjudication number ${this.reportNumber}")
```
[the remark this answers is not in the document]
Comment (p-nik) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt]: changed to validation exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt (hunk 3 above)
Comment (p-nik) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/services/reported/PunishmentsService.kt]: check removed. ticket for front end:  https://dsdmoj.atlassian.net/browse/NN-5121

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/PunishmentsController.kt [hunk 5]
```diff
@@ -110,4 +118,37 @@ class PunishmentsController(
   fun getSuspendedPunishments(
     @PathVariable(name = "prisonerNumber") prisonerNumber: String,
   ): List<SuspendedPunishmentDto> = punishmentsService.getSuspendedPunishments(prisonerNumber = prisonerNumber)
+
+  @Operation(
```
[the remark this answers is not in the document]
Comment (p-nik) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsmanageadjudicationsapi/controllers/reported/PunishmentsController.kt:122]: tests added.

**Review by timoj58 — APPROVED** [review]

**Pull request merged**